Update __init__.py

glenn-sorrentino opened the pull request:
Include "Greenhouse Project" in list of article types.

brassy-endomorph commented:
We probably want to just add a CLI tool or web UI route for adding article types. And in fact we probably eventually want to remove the autopopulation on the assumption that a random user of this as a FOSS tool wouldn't want our article types. I can direct you on that if you want.

glenn-sorrentino commented:
> We probably want to just add a CLI tool or web UI route for adding article types. And in fact we probably eventually want to remove the autopopulation on the assumption that a random user of this as a FOSS tool wouldn't want our article types. I can direct you on that if you want.

Added a CLI tool to create new article types. I rebased, and it added a few files that already exist in the main repo.

brassy-endomorph commented on frontpage/add_type.py:
```diff
@@ -0,0 +1,33 @@
+import sys
+
+# Ensure there's at least one command-line argument (excluding the script name)
+if len(sys.argv) < 2:
```
This should be added using `click` and not as a standalone script.

glenn-sorrentino replied:
- added `click` and a `cli.py` file. 
- updated `__init__.py`

brassy-endomorph commented on frontpage/add_type.py:
```diff
@@ -0,0 +1,33 @@
+article_type_name = " ".join(sys.argv[1:])
+
+# Connect to SQLite database in the /instance directory
+conn = sqlite3.connect("/instance/your_database.db")
```
If you use `click`, you can just import `frontpage.db.db` and use that without having to manually create a connection. It will then accept the same connection string used for the rest of the app. As configured here, this is error prone.

glenn-sorrentino replied:
👍

brassy-endomorph commented on frontpage/add_type.py:
```diff
@@ -0,0 +1,33 @@
+c = conn.cursor()
+
+# Create the 'article_types' table if it doesn't exist
+c.execute(
```
Scripts like this shouldn't execute DDL. If the table doesn't exist, something is critically wrong, and the exception should be raise.

glenn-sorrentino replied:
Centralized the functionality into `__init__.pt`

brassy-endomorph commented on frontpage/add_type.py:
```diff
@@ -0,0 +1,33 @@
+
+# Try to insert the new article type, handle potential errors (e.g., duplicate types)
+try:
+    c.execute("INSERT INTO article_types (type_name) VALUES (?)", (article_type_name,))
```
Since we use SqlAlchemy, we should use the models here.

glenn-sorrentino replied:
👍

brassy-endomorph commented on frontpage/add_type.py:
```diff
@@ -0,0 +1,33 @@
+try:
+    c.execute("INSERT INTO article_types (type_name) VALUES (?)", (article_type_name,))
+    conn.commit()
+    print(f"Article type '{article_type_name}' has been added successfully.")
```
If you're single quoting the variable to differentiate it from the rest of the text, you should use the `repr` format like: `"Article type {article_type_name!r} ..."`

glenn-sorrentino replied:
👍

brassy-endomorph commented on frontpage/add_type.py:
```diff
@@ -0,0 +1,33 @@
+    c.execute("INSERT INTO article_types (type_name) VALUES (?)", (article_type_name,))
+    conn.commit()
+    print(f"Article type '{article_type_name}' has been added successfully.")
+except sqlite3.IntegrityError as e:
```
Catching the script prevents the exception from being raised which means the script will exit 0 on error. This is undesirable generally. It's better to just let the error fully surface for now so we get a stacktrace. If we very carefully caught the error on exactly the unique key's constraint, then printing this message and exiting 1 would be appropriate, but as this is, it will hide potential bugs.

glenn-sorrentino replied:
👍

brassy-endomorph commented on frontpage/routes.py:
```diff
@@ -306,7 +306,7 @@ def article(slug):
     # Convert the download size to a human-readable format for the template
     download_size_formatted = (
-        format_size(int(article.download_size)) if article.download_size is not None else None
+        format_size(int(article.download_size)) if article.download_size else None
```
This will need to be rebased away after your other PR is merged

glenn-sorrentino replied:
👍
